File: Core/src/main/java/jokatu/game/player/AbstractPlayerFactory.java
package jokatu.game.player;

import jokatu.game.Game;
import org.jetbrains.annotations.NotNull;

import java.text.MessageFormat;

/**
 * A {@link PlayerFactory} that produces players for a specific game type.
 * @param <P> the type of {@link Player} to produce
 * @param <G> the type of {@link Game} to create the players for
 */
public abstract class AbstractPlayerFactory<P extends Player, G extends Game<P>> implements PlayerFactory<P> {

	@NotNull
	protected abstract Class<G> getGameClass();

	@NotNull
	private G castGame(@NotNull Game<?> game) {
		Class<G> gameClass = getGameClass();
		if (!gameClass.isInstance(game)) {
			throw new IllegalArgumentException(
					MessageFormat.format("Game was not of the right type.  Expected {0}.", gameClass)
			);
		}
		return gameClass.cast(game);
	}

	@Override
	@NotNull
	public final P produce(@NotNull Game<?> game, @NotNull String username) {
		return produceInCastGame(castGame(game), username);
	}

	@NotNull
	protected abstract P produceInCastGame(@NotNull G g, @NotNull String username);
}

File: Core/src/main/kotlin/jokatu/game/player/AbstractPlayerFactory.kt
package jokatu.game.player

import jokatu.game.Game
import kotlin.reflect.KClass
import kotlin.reflect.full.cast

/**
 * A [PlayerFactory] that produces players for a specific game type.
 * @param <P> the type of [Player] to produce
 * @param <G> the type of [Game] to create the players for
 */
abstract class AbstractPlayerFactory<P : Player, G : Game<P>>(private val gameClass: KClass<G>) : PlayerFactory<P> {
	private fun castGame(game: Game<*>): G {
		if (!gameClass.isInstance(game)) {
			throw IllegalArgumentException(
					"Game was not of the right type.  Expected ${gameClass.simpleName} but was " +
							"${game.javaClass.simpleName}."
			)
		}
		return gameClass.cast(game)
	}

	override fun produce(game: Game<*>, username: String): P {
		return produceInCastGame(castGame(game), username)
	}

	protected abstract fun produceInCastGame(g: G, username: String): P
}
